rustworkx-core/src/connectivity/core_number.rs

```diff
@@ -0,0 +1,206 @@
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations
+// under the License.
+
+use crate::dictmap::*;
+use hashbrown::{HashMap, HashSet};
+use petgraph::visit::{GraphBase, IntoNeighborsDirected, IntoNodeIdentifiers, NodeCount};
+use petgraph::Direction::{Incoming, Outgoing};
+use std::hash::Hash;
+
+/// Return the core number for each node in the graph.
+///
+/// A k-core is a maximal subgraph that contains nodes of degree k or more.
+///
+/// The function implicitly assumes that there are no parallel edges
+/// or self loops. It may produce incorrect/unexpected results if the
+/// input graph has self loops or parallel edges.
```

Comment:
It would be good if we detected those cases and returned `None`

Reply:
So no Err required? Just change return to `Option `and make it `None` if detected?

Reply:
It looks like there are 5 different modules where that 'implicitly assumes' appears. In #784, I'm creating a `utils.rs` at the `rustworkx-core/src` level. Does it make sense to create a `no_self_loop_parallel` function there that can be called from these  different modules? I'd add a call to it from `core_number` here and do a new issue/PR to add calls from the others.

Reply:
We can handle this in the next release then. It is worth creating an issue just for it.

Reply:
Also, parallel edges have different meanings… for k-core it’s more complicated because there is more than one useful way to handle them